src/placeos-frontend-loader/api/remote/githubremote.cr

```diff
@@ -0,0 +1,176 @@
+      parsed = JSON.parse(response.body).as_a
+      parsed.each do |value|
+        tags << value["tag_name"].as_s
+      end
+      tags
+    end
+
+    def url(repo_name : String) : String
+      "https://github.com/#{repo_name}"
+    end
+
+    def download(
+      ref : Remote::Reference,
+      branch : String? = "master",
+      hash : String? = "HEAD",
+      tag : String? = "latest"
+    )
+      repository_uri = url(ref.repo_name)
+      repository_folder_name = ref.repo_path.split("/").last
+
+      if hash == "HEAD" || hash.nil?
+        if tag != "latest"
+          hash = get_hash_by_tag(repository_uri, tag)
+        elsif branch != "master"
+          hash = get_hash_by_branch(repository_uri, branch)
+        else
+          hash = get_hash_head(repository_uri)
+        end
+        hash = hash.not_nil!
+      end
+
+      Git.repository_lock(repository_folder_name).write do
+        Log.info { {
+          message:    "downloading repository",
+          repository: repository_folder_name,
+          branch:     branch,
+          uri:        repository_uri,
+        } }
+
+        begin
+          archive_url = "https://github.com/#{ref.repo_name}/archive/#{hash}.tar.gz"
+          download_archive(archive_url)
+          extract_archive(ref.repo_path)
+          save_metadata(ref.repo_path, hash, repository_uri, branch)
+        rescue ex : KeyError | File::Error
+          Log.error(exception: ex) { "Could not download repository: #{ex.message}" }
+        end
+      end
+    end
+
+    private def get_hashes(repo_url : String)
+      stdout = IO::Memory.new
+      Process.new("git", ["ls-remote", repo_url], output: stdout).wait
+      output = stdout.to_s.split('\n')
+      output.compact_map do |ref|
+        next if ref.empty?
+        ref.split('\t', limit: 2).reverse
+      end.to_h
+    end
+
+    private def get_hash_head(repo_url : String)
+      ref_hash = get_hashes(repo_url)
+      ref_hash.has_key?("HEAD") ? ref_hash["HEAD"] : ref_hash.first_key?
+    end
+
+    private def get_hash_by_branch(repo_url : String, branch : String)
+      ref_hash = get_hashes(repo_url)
+      raise KeyError.new("Branch #{branch} does not exist in repo") unless ref_hash.has_key?("refs/heads/#{branch}")
+      ref_hash["refs/heads/#{branch}"]
+    end
+
+    # tag = "1.9.0"
+    private def get_hash_by_tag(repo_url : String, tag : String)
+      ref_hash = get_hashes(repo_url)
+      raise KeyError.new("Tag #{tag} does not exist in repo") unless ref_hash.has_key?("refs/tags/v#{tag}")
+      ref_hash["refs/tags/v#{tag}"]
+    end
+
+    def download_archive(url)
+      HTTP::Client.get(url) do |redirect_response|
+        raise HTTP::Server::ClientError.new("status_code for #{url} was #{redirect_response.status_code}") unless (redirect_response.success? || redirect_response.status_code == 302)
+        HTTP::Client.get(redirect_response.headers["location"]) do |response|
+          File.write(TAR_NAME, response.body_io)
+        end
```

Comment:
Is there always a redirect response here?

Reply:
Yeah seems like it

Reply:
Yeah looks like it